src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs

```diff
@@ -86,11 +86,145 @@ private static IEnumerable<TextEdit> FormatRazor(FormattingContext context, Razo
+            // {
+            //     <div></div>
+            // }
+            else if (node is MarkupBlockSyntax markupBlockNode &&
+                markupBlockNode.Parent is CSharpCodeBlockSyntax cSharpCodeBlock)
+            {
+                var previousChild = cSharpCodeBlock.Children.PreviousSiblingOrSelf(markupBlockNode);
+                var nextChild = cSharpCodeBlock.Children.NextSiblingOrSelf(markupBlockNode);
+
+                codeNode = markupBlockNode;
+                openBraceNode = previousChild;
+                closeBraceNode = nextChild;
+            }
+            // We're looking for a code block like this:
+            //
+            // @{
+            //     var x = 1;
+            // }
+            else if (node is CSharpCodeBlockSyntax expliciteCode &&
+                expliciteCode.Children.First() is CSharpStatementSyntax statement &&
+                statement.Body is CSharpStatementBodySyntax csharpStatementBody)
+            {
+                openBraceNode = csharpStatementBody.OpenBrace;
+                codeNode = csharpStatementBody.CSharpCode;
+                closeBraceNode = csharpStatementBody.CloseBrace;
+
+                additionalIndentation = true;
+            }
+            // @functions
+            // {
+            // }
+            else if (node is CSharpCodeBlockSyntax directiveCode &&
+                directiveCode.Children.Count == 1 && directiveCode.Children.First() is RazorDirectiveSyntax directive &&
+                directive.Body is RazorDirectiveBodySyntax directiveBody &&
+                directiveBody.Keyword.GetContent().Equals("functions"))
+            {
+                var cSharpCode = directiveBody.CSharpCode;
+                var openBrace = cSharpCode.Children.FirstOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode);
+                var closeBrace = cSharpCode.Children.LastOrDefault(c => c.Kind == SyntaxKind.RazorMetaCode);
+                var code = cSharpCode.Children.First(c => c.Kind == SyntaxKind.CSharpCodeBlock) as CSharpCodeBlockSyntax;
+
+                openBraceNode = openBrace;
+                codeNode = code!;
+                closeBraceNode = closeBrace;
+            }
+            else
+            {
+                // This isn't a node we care about
+                return;
+            }
+
+            var openBraceRange = openBraceNode.GetRangeWithoutWhitespace(source);
+            var codeRange = codeNode.GetRangeWithoutWhitespace(source);
+            if (openBraceRange is not null &&
+                codeRange is not null &&
+                openBraceRange.End.Line == codeRange.Start.Line &&
+                // Because we don't always know what kind of Razor object we're operating on we have to do this to avoid duplicate edits.
+                // The other way to accomplish this would be to apply the edits after every node and function, but that's not in scope for my current work.
+                !edits.Any(e => e.Range.End == codeRange.End))
```

Comment:
I don't think I understand this. Why would there be duplicate edits here? And even if there are, this isn't really checking for duplicates, is it? Actual duplicates would be filtered out at the end of formatting, when it resolves the minimal set.

Reply:
The problem here is again that we don't have good ways of knowing which language concepts we're working on (most things are just a `CSharpCodeBlockSyntax`). Since we're guessing there are scenarios for which multiple of these conditions will try to apply to different nodes which represent the same text (or subsections of the same text). This can manifest itself as things like:
`@{ var x = "t";}`
turning into:
```razor
@{

    var x = "t";

}
```
instead of
```razor
@{
    var x = "t";
}
```
Does that make any sense?

Reply:
> we're guessing there are scenarios for which multiple of these conditions will try to apply to different nodes

That seems to me like the conditions being too broad. Is there anything we can do to make the conditions more strict, or is this all we can do until the parser improves? I was hoping that one of the big wins of moving to the syntax tree based model was the removal of guesses, but maybe thats too pie-in-the-sky?

Reply:
This is as narrow as I could figure out how to make our conditions until the parser improves. I think we DID get value in avoiding some guesses, but until the SyntaxTree can be more informative some of these guesses seem unavoidable.